src/System.Text.Json/src/System/Text/Json/JsonEncodedText.cs

```diff
@@ -0,0 +1,148 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers;
+using System.Diagnostics;
+
+namespace System.Text.Json
+{
+    public readonly struct JsonEncodedText : IEquatable<JsonEncodedText>
+    {
+        private readonly byte[] _utf8Value;
+        private readonly string _value;
+
+        public ReadOnlySpan<byte> EncodedUtf8Bytes => _utf8Value;
+
+        public static JsonEncodedText Encode(string value)
+        {
+            if (value == null)
+                throw new ArgumentNullException(nameof(value));
+
+            return Encode(value.AsSpan());
+        }
+
+        public static JsonEncodedText Encode(ReadOnlySpan<char> value)
+        {
+            if (value.Length == 0)
+            {
+                return new JsonEncodedText(Array.Empty<byte>());
+            }
+
+            return TranscodeAndEncode(value);
+        }
+
+        private static JsonEncodedText TranscodeAndEncode(ReadOnlySpan<char> value)
+        {
+            JsonWriterHelper.ValidateValue(value);
+
+            int expectedByteCount = JsonReaderHelper.GetUtf8ByteCount(value);
+            byte[] utf8Bytes = ArrayPool<byte>.Shared.Rent(expectedByteCount);
+
+            JsonEncodedText encodedText;
+
+            try
+            {
+                // Since GetUtf8ByteCount above already throws on invalid input, the transcoding
+                // to UTF-8 is guaranteed to succeed here. Therefore, there's no need for a catch block.
+                int actualByteCount = JsonReaderHelper.GetUtf8FromText(value, utf8Bytes);
+                Debug.Assert(expectedByteCount == actualByteCount);
+
+                encodedText = EncodeHelper(utf8Bytes.AsSpan(0, actualByteCount));
+            }
+            finally
+            {
+                // On the basis that this is user data, go ahead and clear it.
+                utf8Bytes.AsSpan(0, expectedByteCount).Clear();
+                ArrayPool<byte>.Shared.Return(utf8Bytes);
+            }
+
+            return encodedText;
+        }
+
+        public static JsonEncodedText Encode(ReadOnlySpan<byte> utf8Value)
+        {
+            if (utf8Value.Length == 0)
+            {
+                return new JsonEncodedText(Array.Empty<byte>());
+            }
+
+            JsonWriterHelper.ValidateValue(utf8Value);
+            return EncodeHelper(utf8Value);
+        }
+
+        private static JsonEncodedText EncodeHelper(ReadOnlySpan<byte> utf8Value)
+        {
+            int idx = JsonWriterHelper.NeedsEscaping(utf8Value);
+
+            if (idx != -1)
+            {
+                return new JsonEncodedText(GetEscapedString(utf8Value, idx));
+            }
+            else
+            {
+                return new JsonEncodedText(utf8Value.ToArray());
+            }
+        }
+
+        private JsonEncodedText(byte[] utf8Value)
```

Comment:
I'd move this up to between the fields and the factory methods.  It makes it much more code-reviewer visible that there's a private ctor (and no public ctors).